<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <script src="https://d3js.org/d3.v5.min.js"></script>
	<script src="http://dagrejs.github.io/project/graphlib-dot/latest/graphlib-dot.min.js"></script>
	<script src="https://dagrejs.github.io/project/dagre-d3/latest/dagre-d3.min.js"></script>
    <style>
    	#chart {
			position: fixed;
			left: 0px;
			right: 0px;
			top: 0px;
			bottom: 0px;
			z-index: -2;
			height: 100%;
		}
		body {
			font-family: sans-serif
		}
		div.tooltip {
			position: absolute;
			text-align: left;
			left: 5%;
			top: 5%;
			width: 20%;
			height: 15%;
			padding: 2px;
			border: 0px;
			border-radius: 2px;
			pointer-events: none;
		}
		.node circle{
			stroke: #fff;
			stroke-width: 2px;
			fill: #ffd699; 
		}
		.node rect {
			stroke: #fff;
			stroke-width: 2px;
			fill: #ffc2c2; 
			rx: 10;
			ry: 10;
		}
		.edgeLabel rect {
			fill: #fff;
		}
		.edgePath {
			stroke: #8e8e8e;
			stroke-width: 0px;
			fill: none;
		}
        .label {
            font-size: 12px
        }
        div.tooltip {
            position: absolute;
            text-align: left;
            width: 150px;
            height: auto;
            padding-left: 10px;
            padding-bottom: 10px;
            font: 12px sans-serif;
            border: 0.5px solid gray;
            background: white;
            border-radius: 5px;
            pointer-events: none;
            padding-top: 10px;
            box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
        }
    </style>
  </head>
  <body>
	<div>
        Filter Artifacts: <input type="checkbox" id="filterArtifacts">
        <br>
    </div>
    <div id="chart">
    </div>
    <script>
        var chartDiv = document.getElementById("chart");
        var width = chartDiv.clientWidth -40;
        var height = chartDiv.clientHeight -40;

        var x = d3.scaleLinear().range([0,width]).domain([0,1000])
        var y = d3.scaleLinear().range([0,height]).domain([0,1000])
        var scale = d3.scaleLinear().range([0,width]).domain([0,1000])

        var svg = d3.select("body").append("svg")
            .attr("width", width)

        var div = d3.select("body").append("div")
            .attr("class", "tooltip")
            .style("opacity", 0);
		var filterArtifacts = d3.select('#filterArtifacts')
		var render = new dagreD3.render();

		
		
		var dotfile = 'data/lichee.dot'
		var driverGeneFile = 'data/driverGenes.txt'
		var signatureFile = 'data/nodeSignatures.txt'
		var ccfFile = 'data/ccfs.txt'

        var signatureColors = ['#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c','#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffff99','#b15928','#8dd3c7','#ffffb3','#bebada','#fb8072','#80b1d3','#fdb462','#b3de69','#fccde5','#d9d9d9','#bc80bd','#ccebc5','#ffed6f','#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c']
        var signatureDesc = ['Sponatneous deanimation of 5-methylcytosine (clock-like signature)',
                            'Activity of APOBEC family of cytidine deaminases',
                            'Defective homologous recombination DNA damage repair',
                            'Tobacco smoking',
                            'Unknown (clock-like signature)',
                            'Defective DNA mismatch repair',
                            'Ultraviolet light exposure',
                            'Ultraviolet light exposure',
                            'Ultraviolet light exposure',
                            'Ultraviolet light exposure',
                            'Unknown',
                            'Polimerase eta somatic hypermutation activity',
                            'Polymerase epsilon exonuclease domain mutations',
                            'Polymerase epsilon exonuclease domain mutations',
                            'Temozolomide treatment',
                            'Unknown',
                            'Activity of APOBEC family of cytidine deaminases',
                            'Concurrent polymerase epsilon mutation and defective DNA mismatch repair',
                            'Defective DNA mismatch repair',
                            'Unknown',
                            'Unknown',
                            'Unknown',
                            'Damage by reactive oxygen species',
                            'Unknown',
                            'Concurrent POLD1 mutations and defective DNA mismatch repair',
                            'Defective DNA mismatch repair',
                            'Aristolochic acid exposure',
                            'Unknown',
                            'Aflatoxin exposure',
                            'Chemotherapy treatment',
                            'Defective DNA mismatch repair',
                            'Possible sequencing artifact',
                            'Unknown',
                            'Tobacco chewing',
                            'Defective DNA base excision repair due to NTHL1 mutations',
                            'Platinum chemotherapy treatment',
                            'Azathioprine treatment',
                            'Unknown',
                            'Unknown',
                            'Platinum chemotherapy treatment',
                            'Defective DNA base excision repair due to MUTYH mutations',
                            'Unknown',
                            'Indirect effect of ultraviolet light',
                            'Unknown',
                            'Unknown',
                            'Unknown',
                            'Haloalkane exposure',
                            'Possible sequencing artifact',
                            'Defective DNA mismatch repair',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Possible sequencing artifact',
                            'Activity of activation-induced cytidine deminase (AID)',
                            'Indirect effects of activation-induced cytidine deaminase (AID)',
                            'Unknown chemotherapy treatment',
                            'Thiopurine chemotherapy treatment',
                            'Colibactin exposure (E.coli bacteria carrying pks pathogenicity island)',
                            'Unknown',
                            'Duocarmycin exposure']
        
        var signatures = ["SBS1","SBS2","SBS3","SBS4","SBS5","SBS6","SBS7a","SBS7b","SBS7c","SBS7d","SBS8","SBS9","SBS10a","SBS10b","SBS11","SBS12","SBS13","SBS14","SBS15","SBS16","SBS17a","SBS17b","SBS18","SBS19","SBS20","SBS21","SBS22","SBS23","SBS24","SBS25","SBS26","SBS27","SBS28","SBS29","SBS30","SBS31","SBS32","SBS33","SBS34","SBS35","SBS36","SBS37","SBS38","SBS39","SBS40","SBS41","SBS42","SBS43","SBS44","SBS45","SBS46","SBS47","SBS48","SBS49","SBS50","SBS51","SBS52","SBS53","SBS54","SBS55","SBS56","SBS57","SBS58","SBS59","SBS60","SBS84","SBS85","SBS86","SBS87","SBS88","SBS89","SBS90"] 
        var all_signatures = [...signatures]
		var removedSignatures = []

        var colorScale = d3.scaleOrdinal().range(signatureColors).domain(all_signatures)
        var descScale = d3.scaleOrdinal().range(signatureDesc).domain(all_signatures)

		d3.text(dotfile).then(function(data){
			customPaths = svg.append('g').attr('class','customPaths')
			dagre = svg.append('g').attr('class','dagre')
			
			data = data.replace(/width=1920 height=1080/g,"") //remove lichee's incorrect width and height attribute
			data = data.replace(/shape=([\S]+)/g,"") //remove lichee's incorrect shape attribute
			//console.log(data)
			var graph = graphlibDot.read(data)
			graph.graph().rankdir="LR"
            graph.graph().ranksep="100"
            graph.graph().nodesep="60"
            graph.graph().marginy="20"
			//how to rank tree: allows 'network-simplex' ,'tight-tree', 'longest-path'
			graph.graph().ranker="tight-tree"
			
			graph.nodes().forEach(function(v) {
				var node = graph.node(v);
				//console.log(node)
				if(!isNaN(node.label)){
					node.shape = "circle"
					node.height = Math.cbrt(node.label)*2
					node.id = v
					//console.log(node.height)
				}
				else {
					node.shape = "rect"
					node.height = 30
				}
			});
			
			console.log(graph._edgeLabels)
			
			for (i in graph._edgeLabels) {
				graph._edgeLabels[i].curve = d3.curveBasis
				//graph._edgeLabels[i].arrowhead = "customPath"
			}
			
			render(d3.select(".dagre"), graph);
			svg.attr("height",graph.graph().height + 100)
			svg.attr("width",graph.graph().width + 500)
			//console.log(graph)
			var nodes = graph._nodes
			var legend = svg.append('g').attr('class','legend').attr('transform','translate(' + (graph.graph().width + 150) + ',0)')

			d3.selectAll(".label-container")
				.attr('r',function(d){	
					if(this.parentElement.id){
						//console.log(nodes[this.parentElement.id])
						return Math.cbrt(nodes[this.parentElement.id].label)*2
					}
					else{
						return 5
					}
					
				})
			
			d3.selectAll('.node')
				.append('text')
				.attr('id', function(){
					return "genes_" + this.parentElement.id 
				})
				.attr('transform',function(){
					if(this.parentElement.id){
						//console.log(nodes[this.parentElement.id])
						return 'translate(' + (Math.cbrt(nodes[this.parentElement.id].label)*2 + 20) + ",0)"
					}
					return 'translate(0,0)'
				})
			
			//attach parents to node data structure
			for(i in nodes){
				nodes[i].parents = Object.keys(graph._preds[i])
			}
			
			for(i in nodes){
				var d = nodes[i]
				if(d.parents.length > 0){
					var link = customPaths.append('g')
					for(j in d.parents){
						var parent_id = d.parents[j]
						var paths = link.append("path")
							.attr("class", "link")
							.attr("d", function() {
								if(!isNaN(d.label)){
									nodeSize = Math.cbrt(d.label)*2
								}
								else{
									nodeSize = 5
								}
								if(!isNaN(nodes[parent_id].label)){
									parent_nodeSize = Math.cbrt(nodes[parent_id].label)*2
								}
								else{
									parent_nodeSize = 5
								}

								adjustY = d.y + nodeSize
								adjustParentY = nodes[parent_id].y + parent_nodeSize
								adjustY2 = d.y - nodeSize
								adjustParentY2 = nodes[parent_id].y - parent_nodeSize

								return "M" + (d.x) + "," + (adjustY)
									 + "C" + (d.x + nodes[parent_id].x) / 2 + "," + adjustY
									 + " " + (d.x + nodes[parent_id].x) / 2 + "," + adjustParentY
									 + " " + nodes[parent_id].x + "," + adjustParentY
									 + " L" + nodes[parent_id].x + "," + adjustParentY2
									 + " C" + (d.x + nodes[parent_id].x) / 2 + "," + adjustParentY2
									 + " " + (d.x + nodes[parent_id].x) / 2 + "," + adjustY2
									 + " " + d.x + "," + adjustY2

							})
							.attr('fill','#cce6ff')
							.attr('stroke','#cce6ff')
							.attr('stroke-width',2)
							//.attr('opacity',0.7)
					}
				}
				
			}
			var pie = d3.pie()
				.value(function(d){
					if(d.key != 'Sample'){
						return d.value
					}
				})
				.sort(function(a,b){
					if(!removedSignatures.includes(a.key) && removedSignatures.includes(b.key)){
						return -1
					}
					else if(removedSignatures.includes(a.key) && !removedSignatures.includes(b.key)){
						return 1
					}
					else{
						return 0
					}
				})

			function drawSignatures(signature){
				sig_pie = pie(d3.entries(signature))
				//console.log(sig_pie[0].data.value)
				var sig_donut = d3.select('[id="' + sig_pie[0].data.value + '"]')
					.append('g')
					.attr('class','donut')
					
				sig_donut.selectAll()
					.data(sig_pie)
					.enter()
					.append('path')
					.attr('class', function(d){return d.data.key})
					.attr('d', function(d){
						donutRadius = Math.cbrt(nodes[sig_pie[0].data.value].label)*2
						return d3.arc()
							.innerRadius(donutRadius + 1)
							.outerRadius(donutRadius + 15)(d)
					})
					.attr('fill', function(d){
						return colorScale(d.data.key)
					})
					.attr('stroke','white')
					.attr('stroke-width', "0.5px")
					.on('mouseover', function(d){
						var num = parseFloat(d.data.value)
						div.html(
							"<b>" + d.data.key + "</b>: " + num.toFixed(4) + '<br>' +
							descScale(d.data.key)
						)
						div.style("left", (d3.event.pageX + 10) + "px")
						div.style("top", (d3.event.pageY + 15) + "px")
						div.style('opacity','0.9')
						div.style('background', 'gray')
						d3.select(this)
							.attr('stroke','black')
							.attr('stroke-width', '1px')
					})
					.on('mouseout', function(){
						d3.select(this)
							.attr('stroke','white')
						div.style('opacity','0')
					})
			}

            d3.tsv(signatureFile).then(function(signatures){

				function removeSignatures(sigsToRemove){
					for(j in sigsToRemove){
						if(!removedSignatures.includes(sigsToRemove[j])){
							removedSignatures = removedSignatures.concat(sigsToRemove[j])
							document.getElementById(sigsToRemove[j]+'text').style.fill = 'lightgray'
						}
						else{
							removedSignatures.splice(removedSignatures.indexOf(sigsToRemove[j]),1)
							document.getElementById(sigsToRemove[j]+'text').style.fill = 'black'
						}
					}
					svg.selectAll('.donut').remove()
					for(i in signatures){
						drawSignatures(signatures[i])
					}
					for(i in removedSignatures){
						d3.selectAll('.' + removedSignatures[i])
							.attr('fill','none')
							.attr('stroke','none')
					}
					
				}

                console.log(signatures)
                for(i in signatures){
					drawSignatures(signatures[i])
                }

				filterArtifacts.on('click', function(d){
					artifactSignatures = ['SBS27','SBS43','SBS45','SBS46','SBS47','SBS48','SBS49','SBS50','SBS51','SBS52','SBS53','SBS54','SBS55','SBS56','SBS57','SBS58','SBS59','SBS60']
					removeSignatures(artifactSignatures)
				})
				for(i in all_signatures){
					legend.append('rect')
						.attr('x', -100)
						.attr('y', i*12)
						.attr('fill', function(d){
							return colorScale(all_signatures[i])
						})
						.attr('height',10)
						.attr('width',10)
						.attr('id', all_signatures[i] + 'rect')
						.on('mouseover', function(){
							div.html(
								"<b>" + (this.id).slice(0,-4) + "</b>: <br>"+
								descScale(this.id.slice(0,-4)) + "<br><br> (click to hide)"
							)
							div.style("left", (d3.event.pageX + 10) + "px")
							div.style("top", (d3.event.pageY + 15) + "px")	
							div.style('opacity','0.9')
							div.style('background', 'gray')
							d3.select(this)
								.attr('stroke','black')
								.attr('stroke-width', '2px')
						})
						.on('mouseout', function(){
							d3.select(this)
								.attr('stroke','none')
							div.style('opacity','0')
						})
						.on('click', function(){
							removeSignatures([this.id.slice(0,-4)])
						})

					legend.append('text')
						.attr('x', -90)
						.attr('y', i*12)
						.attr('id', all_signatures[i] + 'text')
						.text(all_signatures[i])
						.attr('fill','black')
						.attr('dy',"10px")
						.style('font-size', "12px")
				}
            })

            d3.text(ccfFile).then(function(ccf){
				var headers = ["node_id","ccf"].join("\t");
				var nodeCCF = d3.tsvParse(headers + "\n" + ccf)
                console.log(nodeCCF)
				for(i in nodeCCF){
					//var annot_node = graph.node(driverGenes[i].node_id)
					d3.select('#genes_' + nodeCCF[i].node_id)
						.append('tspan')
						.attr('text-anchor','start')
						.attr('dy','1em')
						.attr('x','3')
						.attr('font-size',10)
						.attr('fill', 'blue')
						.text(function(d){
                            //console.log(nodeCCF[i].ccf)
							return nodeCCF[i].ccf
						})
				}
			})
            d3.text(driverGeneFile).then(function(annot){
				var headers = ["gene","node_id","prevalence"].join("\t");
				var driverGenes = d3.tsvParse(headers + "\n" + annot)
				for(i in driverGenes){
					var annot_node = graph.node(driverGenes[i].node_id)
					d3.select('#genes_' + driverGenes[i].node_id)
						.append('tspan')
						.attr('text-anchor','start')
						.attr('dy','1em')
						.attr('x','3')
						.attr('font-size',10)
						.attr('fill', 'darkred')
						.text(function(d){
							return driverGenes[i].gene.split('|')[1]
						})
				}
			})
			
		})
    </script>
  </body>
</html>
